075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,490 @@
+
+On each reconciliation, the logic will be like this:
+
+1. The `KafkaClusterCreator` creates the `KafkaCluster` instance and it includes:
+
+   * checking the scale down (if brokers cannot be removed because they host topics' partitions) and reverting replicas count if it fails.
+   * collecting removed (scale down) and added nodes (scale up). Let's call these lists as `removedNodes` and `addedNodes`.
+
+2. The `KafkaReconciler.scaleDown()` scales the cluster down if the check didn't fail, or isn't aware of any scale down at all because it was just reverted within the `KafkaClusterCreator`.
+
+3. The `KafkaReconciler.podSet()` scales the cluster up with the newly added brokers being ready, together with all the corresponding services, before starting an auto-rebalancing.
+
+4. The `KafkaAutoRebalancingReconciler.reconcile()` being involved and using the FSM to run rebalancing operations, if requested and in the following order:
+
+   * rebalancing on scale down
+   * rebalancing on scale up
+
+What happens for each reconciliation depends on the auto-rebalancing FSM as well as the `KafkaRebalance` custom resource status.
+
+The `KafkaAutoRebalancingReconciler.reconcile()` loads the `Kafka.status.autoRebalance` content:
+
+* `state`: is the FSM state.
+* `lastTransitionTime`: when the transition to that state happened.
+* `modes`: enabled modes (`remove-brokers` and/or `add-brokers`) with corresponding brokers lists if an auto-rebalancing requested.
+* `error`: error message if a previous auto-rebalancing failed.
+
+The FSM is initialized based on the `state` field.
+
+The `Kafka.status.autoRebalance.modes[add-brokers|remove-brokers].brokers` list is filled or updated based on its current content and what's coming from `removeNodes` and `addedNodes`.
+
+Check the `removeNodes`:
+
+  * if empty
+    * no further action and stay with the current `Kafka.status.autoRebalance.modes[remove-brokers].brokers` list (which could mean being empty/not existing too).
+  * if not empty
+    * update the `Kafka.status.autoRebalance.modes[remove-brokers].brokers` by producing a consistent list with its current content and what is in the `removeNodes` list. This cover the case that while rebalancing scale down running, the user requests an update on an already running one.
+
+Check the `addedNodes`:
+
+  * if empty
+    * no further action and stay with the current `Kafka.status.autoRebalance.modes[add-brokers].brokers` list (which could mean being empty/not existing too)
+  * if not empty
+    * update the `Kafka.status.autoRebalance.modes[add-brokers].brokers` by producing a consistent list with its current content and what is in the `addedNodes` list. This cover the case that while rebalancing scale down running, the user requests a new scale up or an update on an already queued one. Or even the case that while rebalancing scale up running, the user requests an update on an already running one.
```

Comment:
I think this is correct only for the Idle state. For the other states, you should first check the state of the ongoing rebalance and if it is complete, clear the existing add or remove brokers, or? Otherwise you might just add and add more nodes into the lists and never clear them.

Also, do you really combine the lists for the removed nodes? I would assume you always take only the new list as you essentially don't care what cleaned the node, but if it is not in the reverted scaledown nodes, you don't care about it anymore, or?

Reply:
> Well if you are in RebalanceOnScaleDown and the user asked already for scale up, but it's scaling up more

Yes, here you just add the nodes to the list because the list is waiting for scale-up rebalance

> if you are in RebalanceOnScaleUp and the user requested more scale up

I think in this case, ideally you want to check if the ongoing rebelance isn't already complete. If it is, you would remove the already rebalanced nodes from the list and only then add the new nodes. But maybe it is too little gain with too much added complexity?

Reply:
> But maybe it is too little gain with too much added complexity?

Exactly. I see your point but just stopping and requesting a new rebalancing with the updated list is simpler. We should also expect that the new rebalancing is going to be faster because the previously added nodes were already part of a rebalancing before and filled with partitions.